target/linux/mpc85xx/image/p1020.mk

```diff
@@ -1,17 +1,30 @@
 define Device/aerohive_hiveap-330
   DEVICE_VENDOR := Aerohive
   DEVICE_MODEL := HiveAP-330
-  DEVICE_PACKAGES := kmod-tpm-i2c-atmel
+  DEVICE_PACKAGES := kmod-tpm-i2c-atmel kmod-mtd-rw uboot-envtools
```

Comment:
I thought about this a bit more. Unfortunately the OpenWrt builders' way of creating all images at once has the limitation that there's effectively a per-subtarget base-installation, any of the DEVICE_PACKAGES are not included. So the initramfs images on download.openwrt.org would be unable to run the scripts without some more fiddling. (I remember also that initramfs "opkg install $PACKAGE" will complain about "no free space", this could be WAed... but ugly).

hmm

Reply:
Given the common issues among mpc85xx, perhaps we could apply these package additions to all of these boards (or at least to p1020)?

Reply:
OK, I think adding these packages to the entire subtarget and setting up the right sysupgrade environment in 21.02.1 (letting users know to sysupgrade to the most recent 21.02.1 before going any further forward) is the right way to go. Building that now.

Reply:
Update here. 

So, on the current initramfs, you can install kmod-mtd-rw and the uboot-envtools just with opkg and a internet connection. There's no "no free space" error anymore. :)

Reply:
I still want to keep these two packages on the entire subtarget as I did in a more recent push. I think they would be very helpful for other boards as we navigate u-boot issues, and they require almost no space (2.4K for kmod-mtd-rw, 16K for uboot-envtools; only kernel and libc dependencies for each).

Reply:
uboot-envtools? yes, that can be added to the mpc85xx target - no problem. But please look around, nobody dares to include kmod-mtd-rw in any targets or sub-targets. What one can find however are mentions of kmod-mtd-rw in the commit messages (and wiki). Now, I know about your MR12 and MR16 commits: 55d2db0e8cebca5aa9739c4db875ef57c4643f6d, af9dee336dacbc61e056bce0c0dbf508dd7453a2 . And these look just right! There was no need to do it differently. 

As for uboot issues: The device previously used zImage. After some hours of testing, it does look like this is broken, which was why I was getting the AP350. It should be working like on ARM and other archs. I wonder why this wasn't reported yet, because I think this is an upstream problem that needs some attention. As zImage would have its own gzip or xz-decompressors the kernel would be well below the 8MiB.

However, I do like the repartitioning and we should stick with it. it does away with the padding in the ~43MiB sysupgrade file openwrt currently ships for the device. This makes more than up for the fact that uboot needs to be patched.

(Also for some reason the device seems to be able to pull a older kernel from "nowhere" when it fails?!. 
The original commit lists a JS28F512M29EWH as the flash chip. If this is true, the thing has 128MiB of flash!)

Reply:
Oh dear. You are right; defaulting kmod-mtd-rw would not be a good choice for reasons you've pointed out. The motivation is different than that for uboot-envtools, since we're using 'safe' tricks there (and not overwriting the bootloader itself).

JS28F512M29EWH is actually a 512Mbit device (see the name).

Reply:
Derp. Do you know what I just noticed? I was looking for kmod-mtd-rw source. But It's not in openwrt.git/packages/kernel/ directory. Instead it lives in the [package feed](https://git.openwrt.org/?p=feed/packages.git;a=blob;f=kernel/mtd-rw/Makefile;h=5cf1e5f42171a1a6a755d0d2d9ac8787996c8273;hb=HEAD). 

From what I can tell, this means that the snapshot initramfs will not include it, even if it was added to the target's main makefile.

Reply:
And another update that included the promised MultiImage initramfs. No more extra fdt to deal with.

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=97607d5a4ed8f9d64a2db9bcddbb95718c466181

I also fixed the RGB leds and did some other stuff.

Reply:
Honestly, this looks neater. I'd really love a way to include full code snippets in sysupgrade metadata, but perhaps today isn't the day.

I appreciate you taking the time to create the device page on the wiki.

I will test tonight and then give you my Tested-by and Signed-off-by to your version.

Reply:
> Honestly, this looks neater. I'd really love a way to include full code snippets in sysupgrade metadata, but perhaps today isn't the day.

ok, I'll make a neater version ;). I do need to work on that message, I know to be careful with it since sysupgrade messages in LuCI can be badly formatted.  

(I also noticed I left the fdt.bin in, this can be removed)

> I will test tonight and then give you my Tested-by and Signed-off-by to your version.

Reply:
I've made some screenshot from luci's upgrade.

https://gist.github.com/chunkeey/fc7be150abb09120c47c2d63bd99414b#file-screenshot-2022-02-16-at-23-17-49-openwrt-backup-flash-firmware-luci-png

https://gist.github.com/chunkeey/fc7be150abb09120c47c2d63bd99414b#file-screenshot-2022-02-16-at-23-11-30-openwrt-backup-flash-firmware-luci-png

Both look terrible, it seems luci doesn't care about the effort we put into the formatting  :(.